### DISCUSSION TITLE
mypy and tempfile module

### DISCUSSION THREAD

**Discussion by alanwilter** [python/typing#1101] [Reviews]
I have this test file `tt.py`:

```python
import os
import tempfile
empty_file = os.path.join(tempfile._get_default_tempdir(), next(tempfile._get_candidate_names()))
```

When I do `mypy` I got these errors:
```bash
$ mypy tt.py
tt.py:3: error: Module has no attribute "_get_default_tempdir"
tt.py:3: error: Module has no attribute "_get_candidate_names"
Found 2 errors in 1 file (checked 1 source file)
```

How to solve them? I didn't find any `stub` for `tempfile` so far.

**Comment by Akuli** [comment]
The correct solution is likely using public functions instead, e.g. `tempfile.gettempdir()`, `tempfile.mktemp()`, `tempfile.TemporaryFile()`. But if you are sure that these functions don't work for your use case, and you really need `_get_default_tempdir` and `_get_candidate_names`, you can make a pull request to typeshed that adds them to [stdlib/tempfile.pyi](https://github.com/python/typeshed/blob/master/stdlib/tempfile.pyi).

**Reply by alanwilter** [reply]
Indeed, `tempfile.mktemp()` is enough for me. Thanks!

**Reply by Akuli** [reply]
Note that two programs can end up choosing the same name if you use `tempfile.mktemp()`. You can use `tempfile.mkstemp()` to choose a name and create the file, which fixes this, because once the file exists, the same filename won't be reused for a different temporary file. Or even better, you can use `tempfile.NamedTemporaryFile()` which closes and optionally deletes the tempfile for you :)

**Reply by alanwilter** [reply]
I'm aware of these points, hence my first approach, but in my case, I just want a name (a dummy) for creating a empty file in a S3 bucket exclusive folder.
